homeassistant/components/mqtt/binary_sensor.py

```diff
@@ -180,7 +180,7 @@ def off_delay_listener(now: datetime) -> None:
 
         @callback
         @log_messages(self.hass, self.entity_id)
-        @write_state_on_attr_change(self, {"_attr_is_on"})
+        @write_state_on_attr_change(self, {"_attr_is_on", "_expired"})
```

Comment:
When a `binary_sensor` has `expire_after` set, it will set `self._expired` to `False`. As long as the sensor is not expired, it will stay `False` and will not change. If the sensor expires, the state is always saved and `self._expired` is set to `True`. When a new value is received, the timer will be reset and `self._expired` will be set to `False` again.